aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,299 @@
+        ('logging_alembic_log_level', 'logging.alembic_loglevel'),
+        ('logging_sqlalchemy_loglevel', 'logging.sqlalchemy_loglevel'),
+        ('logging_circus_log_level', 'logging.circus_loglevel'),
+        ('user_email', 'autofill.user.email'),
+        ('user_first_name', 'autofill.user.first_name'),
+        ('user_last_name', 'autofill.user.last_name'),
+        ('user_institution', 'autofill.user.institution'),
+        ('show_deprecations', 'warnings.showdeprecations'),
+        ('task_retry_down_revision_interval', 'transport.task_retry_initial_interval'),
+        ('task_maximum_attempts', 'transport.task_maximum_attempts'),
+    )
+
+    def upgrade(self, config: ConfigType) -> None:
+        for current, new in self.conversions:
+            # replace in profile options
+            for profile in config.get('profiles', {}).values():
+                if current in profile.get('options', {}):
+                    profile['options'][new] = profile['options'].pop(current)
+            # replace in global options
+            if current in config.get('options', {}):
+                config['options'][new] = config['options'].pop(current)
+
+    def downgrade(self, config: ConfigType) -> None:
+        for current, new in self.conversions:
+            # replace in profile options
+            for profile in config.get('profiles', {}).values():
+                if new in profile.get('options', {}):
+                    profile['options'][current] = profile['options'].pop(new)
+            # replace in global options
+            if new in config.get('options', {}):
+                config['options'][current] = config['options'].pop(new)
+
+
+_MIGRATION_LOOKUPS = (Initial, AddProfileUuid, SimplifyDefaultProfiles, AddMessageBroker, SimplifyOptions)
+
+
+def get_current_version(config):
+    """Return the current version of the config.
+
+    :return: current config version or 0 if not defined
+    """
+    return config.get('CONFIG_VERSION', {}).get('CURRENT', 0)
+
+
+def get_oldest_compatible_version(config):
+    """Return the current oldest compatible version of the config.
+
+    :return: current oldest compatible config version or 0 if not defined
+    """
+    return config.get('CONFIG_VERSION', {}).get('OLDEST_COMPATIBLE', 0)
+
+
+def upgrade_config(config: ConfigType, target: int = CURRENT_CONFIG_VERSION) -> ConfigType:
+    """Run the registered configuration migrations up to the target version.
+
+    :param config: the configuration dictionary
+    :return: the migrated configuration dictionary
+    """
+    current = get_current_version(config)
+    used = []
+    while current < target:
+        current = get_current_version(config)
+        try:
+            migrator = next(m for m in _MIGRATION_LOOKUPS if m.down_revision == current)
+        except StopIteration:
+            raise exceptions.ConfigurationError(f'No migration found to upgrade version {current}')
```

Comment:
Think it is important to add a test for the two error conditions (migration not found and circular dependency) in both upgrade and downgrade.

Reply:
done

Reply:
(see tests/manage/configuration/migrations/test_migrations.py)